### PR TITLE
Data Instruction support

### PR DESCRIPTION
**Pull request by wynick27**
Implemented feature mentioned in #632 #716
Currently supports disassemble and assemble integer (1-16 bytes), float (single and double), string (ascii and unicode) and mmx, xmm, ymm data instructions.
80 bit extended float is not supported.
Adds another analysis that can analyze functions, xrefs and data. (analadv command)
Allows code mixed with data.
Currently information is stored in encode map file.
image

---
This change is [Reviewable](https://reviewable.io/reviews/x64dbg/x64dbg/758)

### CODE REVIEW

File: src/dbg/_exports.cpp [hunk 1]
```diff
@@ -1049,6 +1056,41 @@ extern "C" DLL_EXPORT duint _dbg_sendmessage(DBGMSG type, void* param1, void* pa
...
+    {
+        EncodeMapDelSegment((duint)param1);
+    }
+    break;
+
+    case DBG_RELEASE_ENCODE_TYPE_BUFFER:
+    {
+        EncodeMapReleaseBuffer(param1);
+    }
+
```
Comment (mrexodia) [inline src/dbg/_exports.cpp]: this will fall through to `DBG_GET_STRING_AT` right?

File: src/dbg/advancedanalysis.cpp [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            FunctionDelete(end);
+            FunctionAdd(start, end, false, icount);
+        }
+    }
+    GuiUpdateAllViews();
+}
+
+void AdvancedAnalysis::analyzeFunction(duint entryPoint, bool writedata)
+{
+    //BFS through the disassembly starting at entryPoint
```
Comment (mrexodia) [inline src/dbg/advancedanalysis.cpp]: Notice that this code does not handle overlapping blocks correctly, not sure if this will cause problems later.

File: src/dbg/advancedanalysis.cpp [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+            byte type = mEncMap[xref.from - mBase];
+            if(desttype == enc_code && type != enc_unknown && type != enc_code)
+                xref.valid = false;
+            else if(desttype == enc_middle)
+                xref.valid = false;
+        }
+    }
+}
+
+bool isFloatInstruction(x86_insn opcode)
```
Comment (mrexodia) [inline src/dbg/advancedanalysis.cpp]: Maybe we should add this to `capstone_wrapper`? Also this list appears to be pretty incomplete, is this just for float references?

File: src/dbg/assemble.cpp [hunk 4]
```diff
@@ -181,8 +191,19 @@ bool assembleat(duint addr, const char* instruction, int* size, char* error, boo
...
     int destSize;
     unsigned char dest[16];
-    if(!assemble(addr, dest, &destSize, instruction, error))
-        return false;
+    unsigned char* newbuffer = nullptr;
+    if(!assemble(addr, dest, 16, &destSize, instruction, error))
+    {
+        if(destSize > 16)
+        {
+            newbuffer = new unsigned char[destSize];  //retry for long string sequence
```
Comment (mrexodia) [inline src/dbg/assemble.cpp]: We have `Memory<unsigned char*>` for this to prevent possible memory leaks.

File: src/dbg/breakpoint.cpp [hunk 5]
```diff
@@ -97,7 +97,8 @@ bool BpNew(duint Address, bool Enable, bool Singleshot, short OldBytes, BP_TYPE
 
 bool BpGet(duint Address, BP_TYPE Type, const char* Name, BREAKPOINT* Bp)
 {
-    ASSERT_DEBUGGING("Export call");
+    if(!DbgIsDebugging())
```
Comment (mrexodia) [inline src/dbg/breakpoint.cpp]: Why is this here? Is BpGet called when not debugging? The assert should take care of that (although I never run in debug mode).

**Comment by mrexodia** [comment]
It's looking good so far, but I have to check the code better so I will merge it later.

File: src/dbg/advancedanalysis.cpp (hunk 2 above)
Comment (wynick27) [inline src/dbg/advancedanalysis.cpp]: I know, maybe I'll resolve overlapping problem later. There could be gaps between basic blocks, especially in control flow obfuscated executable.

File: src/dbg/advancedanalysis.cpp (hunk 3 above)
Comment (wynick27) [inline src/dbg/advancedanalysis.cpp]: Yes, it's a work in progress, I plan to add simd instruction as well, but it takes a lot of time to look up the manuals.

File: src/dbg/breakpoint.cpp (hunk 5 above)
Comment (wynick27) [inline src/dbg/breakpoint.cpp]: BpGet  and PatchEnum always get galled when debugging is being stopped.

**Comment by wynick27** [comment]
I fixed the bug in dbg message function and uses Memory template to allocate memory.
I also added temp code list support for encode map, addresses in the list will always be decoded as code and any other address decoding that conflicts with it will be decoded as data.
Currently CIP is added to the list and if instruction in CIP is branch its target is also added.
These screenshots shows the effect of code list.
I write this test program with junk code at rva 0x1a, without analysis, it's wrongly decoded as instruction.
image
After CIP points to that address, it's correctly decoded.
image

**Comment by mrexodia** [comment]
Thanks a lot, I will test the code and commit fixes if necessary...
